client/client.go

```diff
@@ -374,11 +392,25 @@ func (r *NotaryRepository) GetTargetByName(name string) (*Target, error) {
 		return nil, err
 	}
 
-	meta, err := c.TargetMeta(name)
+	if len(roles) == 0 {
+		roles = append(roles, data.CanonicalTargetsRole)
+	}
+	var (
+		meta *data.FileMeta
+	)
+	for i := len(roles) - 1; i >= 0; i-- {
+		meta, err = c.TargetMeta(roles[i], name)
+		if err != nil {
+			// important to error here otherwise there might be malicious
+			// behaviour that prevents a legitimate version of a target
+			// being found
+			return nil, err
+		} else if meta != nil {
+			break
```

Comment:
Non-blocking: it might be slightly easier to read if this just returned the target instead of breaking.  Then we wouldn't have to test for metadata being nil after the loop.